### PR TITLE
Cleaning concurrent write dead code - [MOD-6112]

### PR DESCRIPTION
**Pull request by GuyAv46**
**Describe the changes in the pull request**

This PR Removes dead code related to the concurrent write feature and removes their configurations from the flow-test flow.

1. The `SAFEMODE` option is the default configuration, and setting it was and still does not change anything.
2. The `CONCURRENT_WRITE_MODE` was causing 2 thread pools to be created (with 8 + 20 threads), and an additional thread per index (for its indexer), while they were never used - all their related code was unreachable.

For more information about the investigation, check the [MOD-6112 ticket discussion](https://redislabs.atlassian.net/browse/MOD-6112).

Removing these configurations from the CI flow will speed up our CI

**Main objects this PR modified**
1. Making the `SAFEMODE`, `CONCURRENT_WRITE_MODE`, and `INDEX_THREADS` configurations log that they are deprecated and do nothing.
2. Removing unreachable code
3. Removing the unnecessary configurations from the flow test flow

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

### CODE REVIEW

File: coord/src/module.c [hunk 1]
```diff
@@ -1625,7 +1618,7 @@ static int searchResultReducer(struct MRCtx *mc, int count, MRReply **replies) {
   }
 
   searchRequestCtx_Free(req);
-  RS_CHECK_FUNC(RedisModule_BlockedClientMeasureTimeEnd, bc);
```
Comment (MeirShpilraien) [inline coord/src/module.c:1628]: I guess this change is because we are not supporting older Redis versions anymore?

File: src/concurrent_ctx.c [hunk 2]
```diff
@@ -13,12 +13,9 @@
 
 static arrayof(redisearch_threadpool) threadpools_g = NULL;
 
-int CONCURRENT_POOL_INDEX = -1;
-int CONCURRENT_POOL_SEARCH = -1;
-
 int ConcurrentSearch_CreatePool(int numThreads) {
   if (!threadpools_g) {
-    threadpools_g = array_new(redisearch_threadpool, 4);
+    threadpools_g = array_new(redisearch_threadpool, 1); // Only used by the coordinator, so 1 is enough
```
Comment (MeirShpilraien) [inline src/concurrent_ctx.c:18]: Whats this pool is used for?

File: src/config.c [hunk 3]
```diff
@@ -79,18 +79,20 @@ CONFIG_GETTER(getExtLoad) {
...
 // SAFEMODE
 CONFIG_SETTER(setSafemode) {
-  config->concurrentMode = 0;
+  RedisModule_Log(RSDummyContext, REDISMODULE_LOGLEVEL_WARNING, "SAFEMODE option is deprecated, and has no effect");
   return REDISMODULE_OK;
 }
 
-CONFIG_BOOLEAN_GETTER(getSafemode, concurrentMode, 1)
+// dummy getter for safemode. Assuming that `maxDocTableSize` is never 0, we always return "true"
+CONFIG_BOOLEAN_GETTER(getSafemode, maxDocTableSize, 0)
```
Comment (MeirShpilraien) [inline src/config.c:87]: I do not like this hack with `maxDocTableSize`, lets create another implementation that returns const 1.

File: src/config.c [hunk 4]
```diff
@@ -79,18 +79,20 @@ CONFIG_GETTER(getExtLoad) {
...
 
 CONFIG_SETTER(setConcurentWriteMode) {
-  config->concurrentMode = 1;
+  RedisModule_Log(RSDummyContext, REDISMODULE_LOGLEVEL_WARNING, "CONCURRENT_WRITE_MODE option is deprecated, and has no effect");
   return REDISMODULE_OK;
 }
 
-CONFIG_BOOLEAN_GETTER(getConcurentWriteMode, concurrentMode, 0)
+// dummy getter for CONCURRENT_WRITE_MODE. Assuming that `maxDocTableSize` is never 0, we always return "false"
+CONFIG_BOOLEAN_GETTER(getConcurentWriteMode, maxDocTableSize, 1)
```
Comment (MeirShpilraien) [inline src/config.c:95]: Same comment as above

File: src/document.c [hunk 5]
```diff
@@ -328,44 +317,8 @@ void AddDocumentCtx_Submit(RSAddDocumentCtx *aCtx, RedisSearchCtx *sctx, uint32_
...
-  }
-
-  if (!concurrentSearch) {
-    Document_AddToIndexes(aCtx, sctx);
-  } else {
-    // Deprecated and broken - should pass `DocumentAddCtx` and not `RSAddDocumentCtx`
-    // also, we have to pass a weak ref to the spec, and handle it in the callback.
-    ConcurrentSearch_ThreadPoolRun(threadCallback, aCtx, CONCURRENT_POOL_INDEX);
-  }
+  aCtx->client.sctx = sctx;
```
Comment (MeirShpilraien) [inline src/document.c:320]: I guess we do not need the `aCtx->client.bc` anymore right? Did you deleted it?

**Review by MeirShpilraien — COMMENTED** [review]
:+1:

File: coord/src/module.c (hunk 1 above)
Comment (GuyAv46) [reply coord/src/module.c:1628]: Right. This was added in 6.0.9, and on master, we support only >6 redis

File: src/concurrent_ctx.c (hunk 2 above)
Comment (GuyAv46) [reply src/concurrent_ctx.c:18]: the distributed aggregation coordinator thread pool

File: src/document.c (hunk 5 above)
Comment (GuyAv46) [reply src/document.c:320]: No. Good call

File: src/indexer.c [hunk 6]
```diff
@@ -405,40 +267,10 @@ static void Indexer_Process(DocumentIndexer *indexer, RSAddDocumentCtx *aCtx) {
     }
   }
 
-  int useTermHt = indexer->size > 1 && (aCtx->stateFlags & ACTX_F_TEXTINDEXED) == 0;
```
Comment (GuyAv46) [inline src/indexer.c:408]: `indexer->size` is always 0

File: src/concurrent_ctx.c (hunk 2 above)
Comment (MeirShpilraien) [reply src/concurrent_ctx.c:18]: So why we hard code it to 1?

File: src/indexer.c [hunk 7]
```diff
@@ -393,20 +267,10 @@ static void Indexer_Process(DocumentIndexer *indexer, RSAddDocumentCtx *aCtx) {
...
-      // likely that subsequent entries do indeed have IDs.
-      firstZeroId = NULL;
-    }
-  }
-
   if (!ctx.spec) {
     QueryError_SetCode(&aCtx->status, QUERY_ENOINDEX);
     aCtx->stateFlags |= ACTX_F_ERRORED;
-    goto cleanup;
+    return;
```
Comment (MeirShpilraien) [inline src/indexer.c:273]: No cleanup is required here?

**Review by MeirShpilraien — APPROVED** [review]

**Comment by alonre24** [comment]
Note that the code that is using `CONCURRENT_WRITE_MODE` configuration became unreachable at least 3 years ago here - https://github.com/RediSearch/RediSearch/pull/1257
by setting `aCtx->stateFlags |= ACTX_F_NOBLOCK` hardcoded in `spec.c`

**Pull request merged**